feat(rollup-relayer & gas-oracle): graceful restart

colinlyguo opened the pull request:
### Purpose or design rationale of this PR

Doc: https://www.notion.so/scrollzkp/Graceful-restart-of-rollup-relayer-and-gas-oracle-cf194c09c06c44f7a2215f7d4a36a4d8

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

Commit:
feat(rollup-relayer & gas-oracle): graceful restart

Commit:
fix go import and unit tests

Commit:
fix gorm id field

georgehao commented on rollup/internal/orm/transaction.go:
```diff
@@ -0,0 +1,114 @@
+
+// InsertTransaction creates a new transaction record and stores it in the database.
+func (t *Transaction) InsertTransaction(ctx context.Context, contextID string, senderMeta *SenderMeta, tx *gethTypes.Transaction, submitAt uint64) error {
+	rlp := new(bytes.Buffer)
+	if err := tx.EncodeRLP(rlp); err != nil {
+		return fmt.Errorf("failed to encode rlp, err: %w", err)
+	}
```
Seems don't need this

colinlyguo replied:
Need to use `RLPEncoding: rlp.Bytes(),` to store the content of the transaction.

georgehao replied:
oh, i misread the var

georgehao commented on database/migrate/migrations/00015_transaction.sql:
```diff
@@ -0,0 +1,46 @@
+    gas_price           NUMERIC      NOT NULL,
+    gas_limit           BIGINT       NOT NULL,
+    nonce               BIGINT       NOT NULL,
+    submit_at           BIGINT       NOT NULL,           
```
how about use `TIMESTAMP`

georgehao commented on database/migrate/migrations/00015_transaction.sql:
```diff
@@ -0,0 +1,46 @@
+    gas_price           NUMERIC      NOT NULL,
+    gas_limit           BIGINT       NOT NULL,
+    nonce               BIGINT       NOT NULL,
+    submit_at           BIGINT       NOT NULL,           
```
this is confused, in general `xx_at` is a timestamp. So I think can change to `submit_block_number`

georgehao replied:
and reserve the submit_at to record the submit time

colinlyguo replied:
changed in [88de67b](https://github.com/scroll-tech/scroll/pull/1076/commits/88de67b28bfa9a0183d642c1d6e44a37ffefb732).
and in inserting a new pending transaction, `submit_at` is equal to `created_at`, so we can check the submit time in `created_at`.

georgehao commented on rollup/internal/controller/sender/sender.go:
```diff
@@ -446,7 +397,7 @@ func (s *Sender) resubmitTransaction(feeData *FeeData, auth *bind.TransactOpts,
-func (s *Sender) checkPendingTransaction(header *types.Header, confirmed uint64) {
+func (s *Sender) checkPendingTransaction(header *gethTypes.Header, confirmed uint64) {
 	number := header.Number.Uint64()
 	atomic.StoreUint64(&s.blockNumber, number)
```
If deployed replicas, I think this number is not right.

if receipt on A, but insert on B, the inserted block number will be wrong.

colinlyguo replied:
Yes. `submit_block_number` is not the block number for which the transaction is included. It only reflects the roughly submit time so that the sender can decide whether to resubmit or not.

colinlyguo replied:
If the node's sync heights of each replica are the same (or up to the latest block), it's OK.

But if not, there would be some problem triggering the resubmit logic, either too early or too late.

colinlyguo replied:
fixed in [f22b884](https://github.com/scroll-tech/scroll/pull/1076/commits/f22b884667637d4fd164dca65ea6df6a12c71d76)

georgehao commented on rollup/internal/orm/transaction.go:
```diff
@@ -0,0 +1,114 @@
+func (t *Transaction) UpdateTransactionStatusByContextID(ctx context.Context, contextID string, status types.TxStatus) error {
+	db := t.db.WithContext(ctx)
+	db = db.Model(&Transaction{})
+	db = db.Where("context_id = ?", contextID)
```
I think there need to check the status to prevent repeated update

colinlyguo replied:
Yes, added in [88de67b](https://github.com/scroll-tech/scroll/pull/1076/commits/88de67b28bfa9a0183d642c1d6e44a37ffefb732).

georgehao commented on database/migrate/migrations/00015_transaction.sql:
```diff
@@ -0,0 +1,46 @@
+-- +goose Up
+-- +goose StatementBegin
+
+CREATE TABLE transaction
```
how about change the name to `pending_transaction`?

colinlyguo replied:
changed in [88de67b](https://github.com/scroll-tech/scroll/pull/1076/commits/88de67b28bfa9a0183d642c1d6e44a37ffefb732)

colinlyguo committed:
Merge branch 'develop' into feat-rollup-relayer-gas-oracle-graceful-restart

colinlyguo committed:
chore: auto version bump [bot]

Commit:
address comments

Commit:
fix checkPendingTransaction logic

Commit:
tweak comments

Commit:
remove sender's blockNumber and baseFeePerGas

Commit:
increase coverage

Commit:
add orm test

colinlyguo committed:
Merge branch 'develop' into feat-rollup-relayer-gas-oracle-graceful-restart

zimpha commented on rollup/internal/controller/sender/sender.go:
```diff
@@ -446,141 +384,148 @@ func (s *Sender) resubmitTransaction(feeData *FeeData, auth *bind.TransactOpts,
-						// The receipt can be nil since the confirmed transaction may not be the latest one.
-						// We just ignore it, the caller of the sender pool should handle this situation.
-						log.Warn("Pending transaction is confirmed by one of the replaced transactions", "key", key, "signer", pending.signer.From, "nonce", pending.tx.Nonce())
+			if newTx, err := s.resubmitTransaction(s.auth, tx, baseFee); err != nil {
```
one possible error is that: `one of the old transactions is confirmed` and in such case we should mark the transaction confirmed. But it rarely happens, we can leave it manually.

See the comments of the original function for more context.

colinlyguo replied:
Yes. I use the following logic to handle this case in the loop:
1. If the tx is` confirmed`, update current tx's status as `confirmed`, and the sender's other txs (of the same `nonce`) as `failed`.
https://github.com/scroll-tech/scroll/blob/442c296ec5e3c5866bbdecc8df53bf9f2ee5d00a/rollup/internal/controller/sender/sender.go#L419-L435
2. When resubmitting txs, check if the "pending" tx has been marked `failed` because it may be confirmed by other "replaced" txs within the same loop.
https://github.com/scroll-tech/scroll/blob/442c296ec5e3c5866bbdecc8df53bf9f2ee5d00a/rollup/internal/controller/sender/sender.go#L449-L457

colinlyguo replied:
seems that the above logic still has some problems... I need to separate the loop into **`two loops`**, one for confirm checking, and the other for resubmitting.

colinlyguo replied:
added some unit tests covering this case in [1dd1d64](https://github.com/scroll-tech/scroll/pull/1076/commits/1dd1d64a9eaa376d44e3f79369f9c3cd88f92f67)

Commit:
add testSendAndRetrieveTransaction unit test

Commit:
tweak

Commit:
add unit tests

Commit:
Merge branch 'develop' into feat-rollup-relayer-gas-oracle-graceful-restart

colinlyguo committed:
chore: auto version bump [bot]

Commit:
fix unit test

Commit:
access list bug fix

zzq0826 committed:
Update help copy

This sentence has repetitive words. "Total" is duplicated, one of them should be deleted.

Commit:
tweak

colinlyguo commented:
still testing `access list` feature. don't merge this PR yet.

Commit:
Merge branch 'develop' into feat-rollup-relayer-gas-oracle-graceful-restart

colinlyguo committed:
chore: auto version bump [bot]

Commit:
fix CI

Commit:
fix golint

Commit:
trigger ci

Commit:
set tx hash as unique index

Commit:
tweak

Commit:
tweak

colinlyguo commented:
TODOs after merging this PR:
- [x] sepolia: db migrate up and redeployment.
- [x] sepolia: testing graceful restart.
- [x] sepolia: checking db info correctness and CPU usage.
- [ ] grafana: reconfigured metrics.
- [ ] mainnet: db migrate up and redeployment.

colinlyguo committed:
Merge branch 'develop' into feat-rollup-relayer-gas-oracle-graceful-restart

Commit:
revert alerts related metrics change

Commit:
fix a log bug

Commit:
trigger ci

Commit:
remove some outdated comments

colinlyguo committed:
Merge branch 'develop' into feat-rollup-relayer-gas-oracle-graceful-restart

Commit:
remove unused metrics

Commit:
simplify db and enhance orm_test

Commit:
refine orm

Commit:
tweak

Commit:
enhance sender test

Commit:
tweak

Commit:
fix

0xmountaintop commented on common/types/db.go:
```diff
@@ -230,3 +230,66 @@ func (s RollupStatus) String() string {
+	case SenderTypeL2GasOracle:
+		return "SenderTypeL2GasOracle"
+	default:
+		return fmt.Sprintf("Unknown (%d)", int32(t))
```
```suggestion
		return fmt.Sprintf("Unknown SenderType (%d)", int32(t))
```

0xmountaintop commented on common/types/db.go:
```diff
@@ -230,3 +230,66 @@ func (s RollupStatus) String() string {
+	case TxStatusConfirmedFailed:
+		return "TxStatusConfirmedFailed"
+	default:
+		return fmt.Sprintf("Unknown (%d)", int32(s))
```
```suggestion
		return fmt.Sprintf("Unknown TxStatus (%d)", int32(s))
```

colinlyguo commented on common/types/db_test.go:
```diff
@@ -85,3 +85,243 @@ func TestProvingStatus(t *testing.T) {
+		{
+			"Invalid Value",
+			SenderType(999),
+			"Unknown (999)",
```
```suggestion
			"Unknown SenderType (999)",
```

colinlyguo commented on common/types/db_test.go:
```diff
@@ -85,3 +85,243 @@ func TestProvingStatus(t *testing.T) {
+		{
+			"Invalid Value",
+			TxStatus(999),
+			"Unknown (999)",
```
```suggestion
			"Unknown TxStatus (999)",
```

colinlyguo committed:
Apply suggestions from code review

Co-authored-by: HAOYUatHZ <37070449+HAOYUatHZ@users.noreply.github.com>

Commit:
fix db tests
